Step3 구현 완료

guswns1659 opened the pull request:
## 설계 및 구현
1. Step3 도메인 모델 및 객체 간 협력 설계

[도메인 모델 사진 첨부](https://github.com/guswns1659/java-monster-race/blob/guswns1659/img/DomainModel.jpg)

2. 새로운 책임을 수행하는 객체 추가
- Parser : 입력받은 "이름, 타입"을 String [이름, 타입]으로 변환하는 책임
- MonsterConstructor : Game으로부터 받은 몬스터 profile로 몬스터를 생성하는 책임
- Monster : move()라는 공통된 행동을 하는 객체의 타입. 달리기, 비행, 에스퍼의 인터페이스

## 자가 피드백 
- Step 1~2 때 몬스터에 move()라는 행동을 부여하니, 달리기, 비행, 에스퍼의 move()도 구현하기 편했다. 변경에 잘 대응할 수 있었다. 
- 사용자로부터 입력받는 것과 그 값을 처리하는 과정에서 힘들었다. Scanner가 빈값을 입력받아서 BufferedReader로 교체했다. 왜 빈 값이 입력되는지 공부가 필요하다.
- 몬스터의 이름, 타입을 입력받는 부분에 예외처리를 하지 못했다. 숫자가 입력되면 에러로 처리해야 하는데 문자로 인식되서 작동이 된다. 고민이 필요하다. 

## 질문
- getter 사용 시 get를 붙이는 것보다 없을 때 의미가 명확할 때 get을 없애는 것도 괜찮나요? 메서드는 행동을 표현해야 하는 것도 좋다고 알고 있어서 궁금하네요.
ex) monster.getName() -> monster.name()

guswns1659 committed:
20.02.07) Refactor step 1~2 of monater race with Hounx's feedback

Changing package name to monsterRace. Giving responsibility of tryForMove() to monster.

guswns1659 committed:
20.02.07) Refactor step 1~2 of monster race with Hounx's feedback

Changing package name to monsterRace. Giving responsibility of tryForMove() to monster.

guswns1659 committed:
20.02.07) Create domain model object of step3

Parsar has responsibility parse user input.
Monsters is interface has tryForMove().
Runner, Flight, esper is Monsters type because of having responsibility
of tryForMove().

guswns1659 committed:
20.02.07) Update reponsibilities of Input and Parser

Adding askProfileOfMonster() in Input for asking name and kinds of
monster.
Creating parser has responsibility parse user input.

guswns1659 committed:
20.02.07) Add MonsterConstructor generating monsters

MonsterConstructor generate monsters based on user input.
MonsterConstructor is composite relationship with Game.
MonsterConstructor is part of Game.

guswns1659 committed:
20.02.07) Add tryForMove() in each monster

Each monster override of tryForMove() of interface monsters.
Update print() of Output.

guswns1659 committed:
20.02.07) Add printWinner() in Output.

Comparing Monsters' moveCount to determine winners.

guswns1659 committed:
20.02.08) Refactor step3

Adding Exception handling when Input asked user.
Changing unnecessary member variables to local variables.

ksundong commented on src/monsterRace/Game.java:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    private void generateMonsters(List<String[]> profilesOfMonsters) {
+        this.monsters = monsterConstructor.createAll(profilesOfMonsters);
```
와 이 코드 뭔가 배웁니다! 멋지네요!

guswns1659 replied:
@ksundong 이렇게 리뷰 주시니 감사합니다! 그런데 위 코드에서 구체적으로 어떤 부분이 좋으셨어요??

ksundong replied:
monsterConstructor.createAll하는 부분이요!
뭔가 코드가 어떤 역할을 하는지 딱 읽혀서 좋았어요!

guswns1659 replied:
@ksundong 객체한테 하나의 책임 주려고 노력했는데 가독성을 높인 것 같아 좋네요! 감사합니다. 내 코드도 읽기 힘든데 타인의 코드 코멘트 주시다니!! 💯

ksundong replied:
@guswns1659 읽으면서 배울점도 있고, 한 번 읽으면 그 다음부터는 쭉쭉 읽혀요!
다른 분들 노력하신게 눈에 확 보이구요. 코멘트는 모르겠지만, 한 번씩 읽어보시는거 강추합니다!

ksundong replied:
왜 전 멘션이 안될까요... 오타 때문이었습니다...

SonHyunJun replied:
@ksundong 글쿠나 ㅎㅎ 저도 꾸준히 시도해봐야겠네용 ㅎㅎ

guswns1659 replied:
@ksundong 순간 ㅋㅋ 다른 깃허브 아이디로 코멘트 남겼네요 ㅎㅎ 
넵! 저도 다른 분들 코드 꾸준히 읽어 봐야겠어요!!

ksundong commented on src/monsterRace/Input.java:
```diff
@@ -0,0 +1,68 @@
+
+    private void askProfileOfMonster() throws IOException {
+        System.out.println("경주할 몬스터의 이름과 종류를 입력하세요. ex) 크롱, 달리기");
+        System.out.println("몬스터 종류 : 달리기, 비행, 에스퍼");
```
오 친절한 안내메시지 👍

ksundong commented on src/monsterRace/Parser.java:
```diff
@@ -0,0 +1,7 @@
+
+public class Parser {
+    public String[] parse(String line) {
+        return line.replaceAll(" ", "").split(",");
```
오 저는 split하고 trim을 썼는데, 이렇게도 할 수 있겠네요!

Hyune-c replied:
trim 이랑 replaceAll 은 에러 처리의 범위를 정하는 의도에 따라 틀려질거 같아요.
trim 은 양 끝만 제거하는 반면 replaceAll 은 문장 중간도 제거하니까요.

ksundong reviewed (commented):
코드 잘 봤습니다!!

guswns1659 committed:
20.02.09) Refactor step3

Change enhanced for loop to forEach loop for readability.
Change if statement to switch statement for readability.
Add close method to Input Object after asked user.

guswns1659 committed:
20.02.09) Refactor step3 by removing unnecessary comments

honux77 commented on src/monsterRace/Esper.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Override
+    public String name() {
```
똑같은 코드가 많이 보이는데 인터페이스보다 추상 클래스가 좋지 않았을까요?

guswns1659 replied:
네! 코드 중복이 많아서 추상 클래스로 대체하는 게 나을 것 같아요!

honux77 commented on src/monsterRace/Esper.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    @Override
+    public void move(int attempts) {
```
메소드 오버로딩을 했는데 같은 역할을 수행하지 않고 중첩 호출을 하는데 그것 때문에 코드 이해가 어렵습니다.

honux77 commented on src/monsterRace/Esper.java:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+    private void isMove() {
```
is 메소드는 불리언을 리턴할 것 같은 이름이네요.

honux77 commented on src/monsterRace/Game.java:
```diff
@@ -0,0 +1,56 @@
+    private List<Monster> monsters;
+    private Monster winner;
+    private int attempts;
+    private MonsterConstructor monsterConstructor;
```
생성자가 생각나는 이름이라서 저는 MonsterFactory가 더 좋을 것 같은데 어때요? 더 좋은 이름은 모르겠네요. ㅎㅎ

honux77 commented on src/monsterRace/Game.java:
```diff
@@ -0,0 +1,56 @@
+        this.monsters.forEach(each -> each.move(this.attempts));
+    }
+
+    private void setWinner() {
```
stream max를 써 보면 어떨까요?

honux77 commented on src/monsterRace/Runner.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Override
+    public String type() {
```
type은 enum이 좋지 않을까요? 크게 상관은 없는 듯 해요.

honux77 reviewed (approved):
깔끔하게 잘 짰어요.

honux77 commented:
참 monster.name()나쁘지 않은 것 같아요. 그보다는 getter / setter 자체가 최대한 없는 편이 좋습니다.
